Introduce pluggable secrets provider

Praveen2112 opened the pull request:
This PR introduces a framework for injecting custom secrets provider which allows us to add secrets to  the configurations dynamically. The configurations be resolved are of a pattern like ${resolver_name:key_for_resolving_config_property} - We have also introduced SecretProviderFactory and SecretProvider which allows us to resolve configuration in runtime. We have migrated the environment loader to this framework

Praveen2112 commented:
cc: @electrum

Praveen2112 commented on bootstrap/src/main/java/io/airlift/bootstrap/Bootstrap.java:
```diff
@@ -178,7 +210,7 @@ public Set<String> configure()
         // replace environment variables in property values
         List<Message> errors = new ArrayList<>();
-        properties = replaceEnvironmentVariables(properties, System.getenv(), (key, error) -> {
+        configurationResolver.getResolvedConfiguration(properties, (key, error) -> {
```
@electrum Do we need to migrate the environment config resolution to this framework ?

electrum replied:
That seems like a good idea. It will probably simplify the code and is a good test for the framework.

wendigo commented:
@Praveen2112 can you fix failures?

Praveen2112 commented:
@wendigo Now the CI is green

Praveen2112 commented:
The usage of this PR is shown here - https://github.com/trinodb/trino/pull/22633

wendigo commented:
I'm not super convinced that airlift should have SPI. I think that we could add a way to plug-in an interface implementation into bootstrap like this:

```
Bootstrap app = new Bootstrap(...);

Injector injector = app
                .doNotInitializeLogging()
                .setConfigurationInterpolator(interpolator)
                .setRequiredConfigurationProperties(config)
                .initialize();
```

Where interpolator is:

```
interface ConfigurationInterpolator
{
        Optional<String> interpolate(String namespace, String key);
}
```

Ideally I'd like to provide a pre-configured interpolator to Trino connectors through the `ConnectorContext` that is initialized and injected through the engine.

Praveen2112 commented:
> Ideally I'd like to provide a pre-configured interpolator to Trino connectors through the ConnectorContext that is initialized and injected through the engine.

But how would we make them pluggable or have some custom implementation in the future - Won't we need classloader isolation for its implementation ? Would it be a part of Trino's plugin ? The problem is the scope of this pluggable configuration provider is not specific to Trino connectors and its ecosystem (like password-authenticator or resource-group.manager) but would be applied for `config.properties` for the TrinoServer as well. If we have it as a part of trino spi - then we might need to load the plugins twice - Once during the startup for configuration resolution and then once for normal execution -

wendigo commented:
Yes. And we would add an SPI in Trino

Praveen2112 commented:
Yes and we have a version which adds a SPI in Trino as well - https://github.com/trinodb/trino/pull/22621 - But there were a few questions on how to pass the properties for the same - We had it propagated via `config.properties` - But we might need a separate file for configuring this `ConfigurationResolver` - how to configure them during the startup.

Praveen2112 commented:
I'm fine with both the options - @electrum Can you guide us on which direction to take - We have a skeleton PR which tracks both the approaches - Having the changes in Airlift and having the changes in Trino - Both has its own set of complexity.

electrum commented:
The reason to do it in Airlift via an SPI is to allow pluggable providers with classloader isolation that can handle interpolation for the main server configuration. For example, we'll have a plugin for AWS Secrets Manager that uses the AWS SDK. Using a plugin loader similar to Trino will prevent leaking the plugin's version of the AWS SDK into the main application's classpath.

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/Bootstrap.java:
```diff
@@ -147,6 +159,26 @@ public Set<String> configure()
+            log.info("Loading configuration resolver");
+            Map<String, String> configurationResolverProperties = Collections.emptyMap();
+            String configResolverFile = System.getProperty("configResolver");
+            if (configResolverFile != null) {
```
`configurationResolver` will be null if the config property is not set, which seems wrong

electrum replied:
We'll want to allow configuring multiple providers, as people may have secrets in multiple systems. While this is rare in the common case, it is likely to be useful or necessary during a migration from one system to another.

What do you think about making this file be TOML similar to #1189? Then we could support multiple providers in a single file.

electrum replied:
Another issue... do we need to support environment variable interpolation in the `config_resolver` file? I have a feeling that this will be needed, unfortunately.

Praveen2112 replied:
> We'll want to allow configuring multiple providers, as people may have secrets in multiple systems. While this is rare in the common case, it is likely to be useful or necessary during a migration from one system to another.

We currently handle multiple systems - In the configResolver file we track the list of config providers we need to use. But yes using TOML would make it easy to configure.

Praveen2112 replied:
> configurationResolver will be null if the config property is not set, which seems wrong. 

We would be replacing it with a default - implementation which is based on environment variables -

Praveen2112 replied:
`configuration-resolver.config-files` allow us to inject multiple files into the system and each file could point to a specific config-provider - so we could fetch credentials from multiple system and multiple instances of same system.

Praveen2112 replied:
> What do you think about making this file be TOML similar to https://github.com/airlift/airlift/pull/1189? Then we could support multiple providers in a single file.
We could try - But I saw @martint had a few comments regarding its conversion to `Map<String, String>` - How are we planning to represent them ?

electrum commented on bootstrap/src/main/java/io/airlift/bootstrap/Bootstrap.java:
```diff
@@ -258,6 +290,8 @@ public Injector initialize()
         moduleList.add(Binder::requireExactBindingAnnotations);
 
+        moduleList.add(binder -> binder.bind(ConfigurationResolver.class)
+                .toInstance(requireNonNull(configurationResolver, "configurationResolver is null")));
```
See above. We shouldn't check for null here.

Praveen2112 replied:
I missed an initialization there -

electrum commented on config-spi/src/main/java/io/airlift/spi/ThreadContextClassLoader.java:
```diff
@@ -0,0 +1,34 @@
+
+import java.io.Closeable;
+
+public class ThreadContextClassLoader
```
Make this a package-private class in the `configuration` module

electrum commented on configuration/src/main/java/io/airlift/configuration/EnvironmentVariableConfigurationValueResolver.java:
```diff
@@ -0,0 +1,34 @@
+{
+    @Override
+    public String resolveConfigurationValue(String key)
+            throws InvalidConfigurationResolverException
```
This exception isn't thrown here

electrum commented on config-spi/src/main/java/io/airlift/spi/config/ConfigurationValueResolver.java:
```diff
@@ -0,0 +1,20 @@
+public interface ConfigurationValueResolver
+{
+    String resolveConfigurationValue(String key)
+            throws InvalidConfigurationResolverException;
```
No need to declare unchecked exceptions. Add a Javadoc instead.

electrum commented on config-spi/src/main/java/io/airlift/spi/config/InvalidConfigurationResolverException.java:
```diff
@@ -0,0 +1,28 @@
+ */
+package io.airlift.spi.config;
+
+public class InvalidConfigurationResolverException
```
How about `ConfigurationResolverException`? This sounds like the resolver itself is invalid, but this is simplify a failure to resolve configuration.

wendigo commented:
@electrum yes, but we can add this SPI in Trino rather than in the airlift

electrum commented on configuration/src/main/java/io/airlift/configuration/resolver/ConfigurationResolverManager.java:
```diff
@@ -0,0 +1,200 @@
+        for (File configFile : configFiles) {
+            builder.put(getNameWithoutExtension(configFile.getName()), loadConfigProvider(configFile.getAbsoluteFile()));
+        }
+        builder.put("env", new EnvironmentVariableConfigurationValueResolver());
```
We should consider making `env` be a configurable plugin, just like any other plugin. Some people may want to disable it for security purposes. Or if you're using a different provider, you might not want or need environment variables.

Praveen2112 replied:
But won't it contradict with this comment - https://github.com/airlift/airlift/pull/1217#discussion_r1672323394. We could migrate them to `configuration-resolvers` plugin which captures a set of predefined `ConfigPlugin` -

Praveen2112 replied:
But would it be backward compatible ? Like in order to use the `env` we need to add an entry to the `config-resolver.properties` - Won't it break for the existing Trino deployment

Praveen2112 commented:
@wendigo We have tried it a similar approach (https://github.com/trinodb/trino/pull/22621) but in order to have in Trino we have two main set of issues - 
1. Where would we store the configuration property for this `config_resolver` - It could be a part of `config.properties` - but we need to remove the properties which are used for `config_resolver` - We might need modify the `launcher.py` to add any new additional property file. 
2. Copying the `config-plugin` to a different location during the startup 

In this case we might need some sort of changes in airlift and some in the Trino side. Plus this configuration resolver doesn't fully fit under Plugin lifecycle in case of Trino - These are mostly static during the startup

Praveen2112 commented:
@electrum Thanks a lot for the review. AC

Praveen2112 commented:
@electrum Based on the previous discussion we have migrated the configuration to TOML file and have made the `EnvironmentVariableConfigurationResolver` pluggable (via Config - so it would be used of configured)

wendigo commented:
@Praveen2112 how the TOML is plugged in into existing Configuration binding? Doesn't seem to be a case

Praveen2112 commented:
The toml file path is passed via `configResolver` system property like how we handle for `config.properties`

electrum commented:
Typo "fot" in commit title

Praveen2112 commented:
@electrum Based on our previous conversation - Have migrated the code from Configuration resolver to Secret provider

Praveen2112 commented:
@electrum Gentle ping for this PR

Praveen2112 commented:
@electrum Gentle ping

electrum commented on configuration/src/main/java/io/airlift/configuration/secrets/SecretsResolver.java:
```diff
@@ -0,0 +1,75 @@
+        while (matcher.find()) {
+            String configProviderName = matcher.group(1).toLowerCase(ENGLISH);
+            SecretProvider secretProvider = secretProviders.get(configProviderName);
+            if (secretProvider != null) {
```
This should fail if the secret provider does not exist

electrum commented on configuration/src/main/java/io/airlift/configuration/secrets/SecretsResolver.java:
```diff
@@ -0,0 +1,75 @@
+            String configProviderName = matcher.group(1).toLowerCase(ENGLISH);
+            SecretProvider secretProvider = secretProviders.get(configProviderName);
+            if (secretProvider != null) {
+                String keyName = matcher.group("key");
```
It's strange to mix named and numbered groups. Use names or numbers for both.

electrum commented on configuration/src/main/java/io/airlift/configuration/secrets/env/EnvironmentVariableSecretProvider.java:
```diff
@@ -0,0 +1,33 @@
+    {
+        String envValue = System.getenv().get(key);
+        if (envValue == null) {
+            throw new SecretProviderException(format("References unset environment variable '%s'", key));
```
```java
throw new RuntimeException("Environment variable is not set: " + key);

electrum commented on configuration/src/main/java/io/airlift/configuration/secrets/SecretsResolver.java:
```diff
@@ -0,0 +1,75 @@
+
+public class SecretsResolver
+{
+    private static final Pattern PATTERN = Pattern.compile("\\$\\{([a-zA-Z][a-zA-Z0-9_-]*):(?<key>[^}]+?)}");
```
I'm thinking we should enforce the same naming pattern for keys. We can relax the restriction later if needed, but allowing spaces, control characters or other arbitrary characters seems like a bad idea.

electrum commented on configuration/src/test/java/io/airlift/configuration/secrets/TestSecretsResolver.java:
```diff
@@ -0,0 +1,127 @@
+        SecretsResolver secretsResolver = new SecretsResolver(
+                ImmutableMap.of("prefix", new PrefixedSecretProvider()));
+
+        assertThat(secretsResolver.getResolvedConfiguration(ImmutableMap.of("key", "${unknown_key:key}")))
```
This should fail

electrum commented on secrets-plugin/pom.xml:
```diff
@@ -0,0 +1,92 @@
+    <parent>
+        <groupId>io.airlift</groupId>
+        <artifactId>airlift</artifactId>
+        <version>251-SNAPSHOT</version>
```
This needs to be updated to the latest

electrum commented on secrets-plugin/src/main/java/io/airlift/secrets/keystore/KeystoreSecretProvider.java:
```diff
@@ -0,0 +1,51 @@
+            return readEntity(keyStore, key, keystorePassword);
+        }
+        catch (GeneralSecurityException e) {
+            throw new SecretProviderException(e.getMessage(), e);
```
No need to pass both the message and the cause. Passing the cause will fill in the message.

electrum commented on secrets-plugin/src/main/java/io/airlift/secrets/keystore/KeystoreSecretProviderFactory.java:
```diff
@@ -0,0 +1,57 @@
+    public SecretProvider createSecretProvider(Map<String, String> config)
+    {
+        Bootstrap app = new Bootstrap(
+                new Module() {
```
I think this can be a lambda

electrum commented on secrets-plugin/src/main/java/io/airlift/secrets/keystore/KeystoreUtils.java:
```diff
@@ -0,0 +1,54 @@
+        SecretKeyEntry secretKeyEntry = (SecretKeyEntry) keyStore.getEntry(entityAlias, new KeyStore.PasswordProtection(password.toCharArray()));
+
+        if (secretKeyEntry == null) {
+            throw new SecretProviderException("Key not found '%s'".formatted(entityAlias));
```
```java
throw new RuntimeException("Key not found in keystore: " + entityAlias);

electrum commented on secrets-plugin/src/test/java/io/airlift/secrets/keystore/TestKeystoreSecretProvider.java:
```diff
@@ -0,0 +1,93 @@
+@TestInstance(PER_CLASS)
+final class TestKeystoreSecretProvider
+{
+    private static final String KEYSTORE_PASSWORD = "password";
```
I would inline this

Praveen2112 commented on bootstrap/src/main/java/io/airlift/bootstrap/Bootstrap.java:
```diff
@@ -178,13 +204,10 @@ public Set<String> configure()
-        properties = replaceEnvironmentVariables(properties, System.getenv(), (key, error) -> {
+        properties = ImmutableSortedMap.copyOf(secretsResolver.getResolvedConfiguration(properties, (key, error) -> {
             unusedProperties.remove(key);
             errors.add(new Message(error));
```
I have migrated it to `errors.add(new Message(error.getMessage(), error);`

Praveen2112 commented:
@electrum Thanks a lot for the review - I have addressed most of the comments except 1 - What would be our fallback mechanism if the key for a config provider doesn't exist -
1. We could fail explicitly
2. Skip resolving them.
We could go for option 2 as a config of a similar pattern could be used by other system

Praveen2112 commented on bootstrap/src/test/java/io/airlift/bootstrap/TestBootstrapWithSecretsProvider.java:
```diff
@@ -0,0 +1,140 @@
+
+        Injector injector = bootstrap.initialize();
+
+        assertThat(injector.getInstance(TestConfig.class).getValue()).isEqualTo("${ENV:TEST_KEY}");
```
Or should we escape them ?

Praveen2112 commented:
@electrum Addressed all the comments - Now if the secret provider is not available we would throw an error. We have captured and throw the error via `Consumer`

electrum commented on secrets-spi/src/main/java/io/airlift/spi/secrets/SecretProvider.java:
```diff
@@ -0,0 +1,25 @@
+     * @return resolved value
+     * @throws SecretProviderException if key cannot be resolved.
+     */
+    String generateSecretValue(String key);
```
That sounds good to me

electrum commented on configuration/src/main/java/io/airlift/configuration/TomlConfiguration.java:
```diff
@@ -0,0 +1,88 @@
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+
+public class TomlConfiguration
```
Make this package private for now

Praveen2112 replied:
I'm not sure if we could use it as package private as we use them in `io.airlift.configuration.secrets.SecretsPluginManager`

electrum commented on configuration/src/main/java/io/airlift/configuration/TomlConfiguration.java:
```diff
@@ -0,0 +1,88 @@
+{
+    public static TomlConfiguration createTomlConfiguration(String path)
+    {
+        verify(path.endsWith(".toml"), "TOML files must end with '.toml'");
```
This should use `checkArgument()` since it is validating an argument

electrum commented on configuration/src/main/java/io/airlift/configuration/TomlConfiguration.java:
```diff
@@ -0,0 +1,88 @@
+
+public class TomlConfiguration
+{
+    public static TomlConfiguration createTomlConfiguration(String path)
```
Make this take `File` directly

electrum commented on bootstrap/src/test/java/io/airlift/bootstrap/TestBootstrapWithSecretsProvider.java:
```diff
@@ -0,0 +1,140 @@
+
+        assertThat(injector.getInstance(FooConfig.class).getValue()).isEqualTo("test_value");
+    }
+
```
Remove extra blank line

electrum commented on bootstrap/src/test/java/io/airlift/bootstrap/TestBootstrapWithSecretsProvider.java:
```diff
@@ -0,0 +1,140 @@
+
+        return tomlFile;
+    }
+
```
Remove extra blank line

Praveen2112 commented:
@electrum Except this [comment](https://github.com/airlift/airlift/pull/1217#discussion_r1699300191) all others were addressed

dain commented on secrets-plugin/src/main/java/io/airlift/secrets/keystore/KeystoreUtils.java:
```diff
@@ -0,0 +1,52 @@
+import java.security.KeyStore;
+import java.security.KeyStore.SecretKeyEntry;
+
+public final class KeystoreUtils
```
I don't think we need this utility class.  Let's just move the code to KeystoreSecretProvider

dain commented on secrets-plugin/src/test/java/io/airlift/secrets/keystore/TestKeystoreSecretProvider.java:
```diff
@@ -0,0 +1,91 @@
+    public void setup()
+            throws Exception
+    {
+        keystoreFile = createTempFile(System.getProperty("java.io.tmpdir"), "keystore");
```
Does airlift have a testing temp directory utility like Trino has?

Praveen2112 replied:
I don't think we have any - Even in most on trino we use `deleteOnExit`

Praveen2112 replied:
Or should we use a different API here ?

Praveen2112 replied:
Thank you @electrum !! Migrated to `TestFile`

dain commented on pom.xml:
```diff
@@ -60,6 +60,7 @@
         <module>openmetrics</module>
         <module>packaging</module>
         <module>sample-server</module>
+        <module>secrets-plugin</module>
```
So is the idea that we would put all plugins in one module?  We can always change this later, but I would imagine that for additional plugins we would want separate modules, and if so maybe this should be the secrets-keystore-plugin.

Praveen2112 replied:
No - But for airlift we could try to have all the secrets-plugin in one place -

dain reviewed (commented):
some minor comments but otherwise looks good

Praveen2112 commented on configuration/src/main/java/io/airlift/configuration/secrets/SecretsPluginManager.java:
```diff
@@ -0,0 +1,205 @@
+
+    private void addSecretProviderFactory(SecretProviderFactory secretProviderFactory)
+    {
+        verify(CONFIG_PROVIDER_NAME_PATTERN.matcher(secretProviderFactory.getName()).matches(), "Secret provider name '%s' doesn't match pattern '%s'", secretProviderFactory.getName(), CONFIG_PROVIDER_NAME_PATTERN);
```
But it would break with the `${ENV:...}` which we apply today right ?

dain replied:
Are you sure it would break that.  The matcher is only used to validate the name of the factory. All names are lowered, so I assume this is all case insensitive.

Praveen2112 replied:
Sorry I was confused with the patter which matches the `${ENV:...}` - You are right - we could modify it to work only with lower case

Praveen2112 commented:
@dain Thanks a lot for the review - Except this comment https://github.com/airlift/airlift/pull/1217#discussion_r1699364480 most of them are addressed.

Praveen2112 commented:
@dain All the comments were addressed

Praveen2112 commented:
@dain / @electrum All the comments were addressed.
